staging/src/k8s.io/apiserver/pkg/util/panic/panic.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package panic
+
+import (
+	"fmt"
+	"net/http"
+	"runtime"
+)
+
+// Invoker propagates panic from a spawned goroutine to the request handler
+/* func(w http.ResponseWriter, req *http.Request) {
+        startHandler := make(chan struct{})
+	invoker := NewInvoker()
+	go invoker.InvokeWithPanicProtection(func() {
+		// decide whether we want to run the handler
+                close(startHandler)
+	})
+
+	select {
+	case <-startHandler:
+		invoker.InvokeWithPanicWrapping(func() {
+			handler.ServeHTTP(w, req)
+		})
+	case err := <-invoker.Done():
+		if err != nil {
+			// the spawned goroutine panicked
+			panic(err)
+		}
+	}
+}
+*/
+type Invoker interface {
+	// InvokeWithPanicProtection invokes the given func with panic
+	// protection, if the function panics then:
+	// a) the panic is recovered
+	// b) the stack trace is caprured
+	// c) both the panic value and the stacktrace are captured into an
+	// error object, so it can be propagated to the parent (request
+	// handler goroutine) via a channel
+	// The request handler uses this function to spawn a new goroutine
+	InvokeWithPanicProtection(func())
```

Comment:
The names of the two "Invoke..." methods are so similar that I find it hard to remember which is which. For example, it is not obvious from the names that one creates a separate goroutine and the other does not.

Reply:
renamed to `WrapPanic` in https://github.com/kubernetes/kubernetes/pull/127075